src/main/java/telraam/logic/lapper/slapper/Slapper.java

```diff
@@ -0,0 +1,139 @@
+    private final Logger logger;
+    private final Jdbi jdbi;
+
+    public Slapper(Jdbi jdbi) {
+        this.jdbi = jdbi;
+        this.scheduler = Executors.newScheduledThreadPool(1);
+        this.logger = Logger.getLogger(Slapper.class.getName());
+        this.debounceScheduled = false;
+
+        // Get the lapSourceId, create the source if needed
+        LapSourceDAO lapSourceDAO = jdbi.onDemand(LapSourceDAO.class);
+        if (lapSourceDAO.getByName(SOURCE_NAME).isEmpty()) {
+            lapSourceDAO.insert(new LapSource(SOURCE_NAME));
+        }
+    }
+
+    @Override
+    public void handle(Detection msg) {
+        if (!this.debounceScheduled) {
+            this.debounceScheduled = true;
+            this.scheduler.schedule(() -> {
+                try {
+                    this.calculateLaps();
+                } catch (Exception e) {
+                    logger.severe(e.getMessage());
+                }
+                this.debounceScheduled = false;
+            }, DEBOUNCE_TIMEOUT, TimeUnit.SECONDS);
+        }
+    }
+
+    private void calculateLaps() {
+        logger.info("Slapper: Calculating laps...");
+        this.jdbi.useHandle(handle -> handle.execute(
+                        """
+                                WITH switchovers AS (
+                                	SELECT teamid AS team_id,
+                                		newbatonid,
+                                		timestamp AS start_timestamp,
+                                		COALESCE(
+                                			LEAD(timestamp) OVER (PARTITION BY teamid ORDER BY timestamp),
+                                			timestamp + INTERVAL '1 MONTH'
+                                		) AS next_baton_switch
+                                	FROM batonswitchover
+                                ),
+                                team_detections AS (
+                                	SELECT MIN(station_id) AS station_id,
+                                		MAX(rssi) as rssi,
+                                		date_trunc('second', timestamp) AS timestamp_seconds,
+                                		team_id
+                                	FROM detection d
+                                	LEFT JOIN switchovers s ON d.baton_id = s.newbatonid
+                                		AND d.timestamp BETWEEN s.start_timestamp AND s.next_baton_switch
+                                	WHERE station_id NOT BETWEEN 3 AND 5
+                                		AND rssi > -84
+                                		AND team_id IS NOT NULL
+                                	GROUP BY date_trunc('second', timestamp), team_id
+                                ),
+                                start_times AS (
+                                	SELECT DISTINCT ON (team_id) team_id,
+                                		timestamp_seconds AS start_seconds
+                                	FROM team_detections
+                                	WHERE station_id BETWEEN 2 AND 3
```

Comment:
What's special about stations between 2 and 3 here?

Reply:
> Lets start with the errors for HILOK (1) and VLK (3). If you look at the query , in no_duplicates I make a new column called first_timestamps. This column is aimed to filter out the first registered lap as SL tended to assign one additional lap to most teams. However, this surplus lap primarily came from teams registering some detections at station 7 at the start of the race. Modifying the query to start lap counting only when teams pass station 2 or 3 for the first time solves the lap problem for HILOK and VLK. However it gave teams 9, 12, 16 and 22 an extra lap. After looking at the lap times I couldn't find the cause of this problem as the lap times appeared consistent with those from EL.

https://github.com/12urenloop/Telraam/pull/133#issuecomment-2036716495

Reply:
Ah, okay, thanks for linking, I hadn't figured it out yet